### ISSUE TITLE
Cannot Tab to MenuBar

### ISSUE THREAD

**Issue by tznind** [gui-cs/Terminal.Gui#2404]
**Is your feature request related to a problem? Please describe.**

I have been exploring using a MenuBar as a dropdown menu.  It works like a dropdown style combo box and looks better and gives clearer indication of what the active/new selections will be.

But there is a problem in that the user cannot Tab to the view.  I have tried setting the following but it does not work.

```csharp
CanFocus = true,
TabStop = true
```

![menu-as-combo](https://user-images.githubusercontent.com/31306100/224381320-bf913b61-3974-4a87-8c2f-5290cf6a8a98.gif)

**Describe the solution you'd like**
Tab to work for MenuBar when `TabStop` is set.

**Describe alternatives you've considered**
I can probably still do a workaround by intercepting Tab on the `View` before it  and adding a callback to manually popping the menu open.

I will probably have to do this anyway so that the user can Tab out of the menu and back to other views.

**Comment by tznind** [comment]
So I was experimenting thismorning and it turns out you can tab to (yay!) it but it doesn't open (boo).

I have gotten it to open with:

```csharp
menuBar.Enter += (e) => {
	allowedTypeMenuBar.OpenMenu (0);
};
```
The trouble is that I cannot then close it again!  First I tried these events (which do not get called at all)
```csharp
// Does not fire
dropdownMenu.KeyPress += (k) => {
	if (k.KeyEvent.Key == Key.Tab) {
		dropdownMenu.CloseMenu (false);
	}
};
// Does not fire
dropdownMenu.Leave += (e) => {
	dropdownMenu.CloseMenu (false);
};

```

Next I tried this in the hosting windows `public override bool ProcessHotKey (KeyEvent keyEvent)` which does get fired but doesnt close the menu.  Even though CloseMenu is called! Is there something about manual open the menu (like above) that is not registering events correctly for close @BDisp?

```csharp
if (allowedTypeMenuBar != null &&
		keyEvent.Key == Key.Tab &&
		allowedTypeMenuBar.IsMenuOpen) 
{
		allowedTypeMenuBar.CloseMenu (false, false, false);
}
```

**Comment by tznind** [comment]
Here is a minimum repro

![tabtomenu](https://user-images.githubusercontent.com/31306100/224472812-ae025d03-44fe-4fbf-ab8b-661f724bb61b.gif)


```csharp
using Terminal.Gui;
using System.Data;

Application.Init();

var win = new Window("Example App (Ctrl+Q to quit)");

var mb = new MenuBar(
    new []{new MenuBarItem( new []{
        new MenuItem("Item1",null,()=>{}),
        new MenuItem("Item2",null,()=>{}),
    })}){
    CanFocus = true,
    TabStop = true,
    X = 10,
    Width = 10,
    Text = "Open Me",
    };

mb.Enter += (e) => {
	mb.OpenMenu ();
};

win.Add(new TextField("SomeText"){Width = 9});
win.Add(mb);
Application.Run(win);
Application.Shutdown();
```

**Comment by BDisp** [comment]
@tznind I could make it to open and close. The problem is I have to tab/shift tab twice to close it because when we tab on the TextField, the MenuBar get focus and the Enter event open the menu, where also have a instruction to set focus to the opened menu. Unfortunately since this operation is made during the Enter event for the MenuBar getting focus, on the end is the MenuBar who get the focus. It seems that we can't set a new focus to other view on the Enter or Leave events.

![menu-as-combobox](https://user-images.githubusercontent.com/13117724/225741347-7e851af1-1007-4126-9dcf-d10be3d6bf2b.gif)

**Comment by tig** [comment]
Would it make sense to rename this Issue to "Enable using MenuBar as a replacement for ComboBox?"

**Comment by BDisp** [comment]
> Would it make sense to rename this Issue to "Enable using MenuBar as a replacement for ComboBox?"

I have a WIP where the `ListView` is decouple from the `ComboBox` and will use the `ColectionNavigator`. Instead of "replacement" can I suggest maintaining the current "alternative" and naming the new as `MenuCombo` or `ComboMenu`?

**Comment by tig** [comment]
I agree we should have both. 

I don't believe "combo" as a term makes sense for this one tho. 

Is it really a new view or is it just a MenuBar used in a particular way?

**Comment by BDisp** [comment]
> Is it really a new view or is it just a MenuBar used in a particular way?

About your question must be @tznind who can answer. I only said that I'm improvement the current `ComboBox` and I would like to maintain it, to not be replaced. Then the @tznind idea will be another new view, I think.

**Comment by tznind** [comment]
I agree with @tig  that `MenuBar` should be placeable anywhere and therefore used like a 'Drop Down List' without creating a new class.  This is already working in `FileDialog`.  

So I think we could call this issue done?

**Comment by tig** [comment]
Do the api docs describe how to do this?

**Comment by tznind** [comment]
Just been looking at FileDialog Here is a minimum repro.  I think that what we want is for the only thing the user has to do is set `CanFocus=true`.  The rest (like the `Enter` event) should not be required.


```csharp
Application.Init ();

var w = new Window ();
w.Add (new TextField () { Width = 10 });

var menu = new MenuBarItem ("Dropdown " + Application.Driver.DownArrow,
	Enumerable.Range (1, 5).Select (
			(a) => new MenuItem ("Item " + a.ToString (), null, () => {
				MessageBox.Query ("Clicked", "You clicked " + a, "Ok");
			}))
		.ToArray ());

var mb = new MenuBar (new [] { menu }) {
	CanFocus = true,
	Width = 10,
	Y = 1
};

// HACKS required to make this work:
mb.Enter += (s, e) => {
	// BUG: This does not select menu item 0
	// Instead what happens is the first keystroke the user presses
	// gets swallowed and focus is moved to 0.  Result is that you have
	// to press down arrow twice to select first menu item and/or have to
	// press Tab twice to move focus back to TextField
	mb.OpenMenu ();
};
			
w.Add (mb);

Application.Run (w);

Application.Shutdown ();
```

**Comment by tig** [comment]
Why isn't Onfocus true by default?

Yesterday I was goofing with Focus stuff and realized it might make sense for Menubar to participate in tabstops along with the other subviews of a view.

**Comment by BDisp** [comment]
That's not the normal behavior on a app, but you can do that behavior for `Terminal.Gui`. It's debatable.

**Comment by tig** [comment]
> That's not the normal behavior on a app, but you can do that behavior for `Terminal.Gui`. It's debatable.

Yeah. Probably best to leave it as is. It was just an idle thought.

**Comment by tig** [comment]
Here's a tweaked scenario.  I realy love this!

```cs
		public override void Setup ()
		{
			var tf = new TextField () { Width = 10 };
			Application.Top.Add (tf);

			MenuBarItem menu = null;

			MenuItem CreateMenuItem (int i)
			{
				return new MenuItem ($"Item {i}", null, () => {
					tf.Text = menu.Children [i - 1].Title;
				});
			}

			menu = new MenuBarItem ($"{Application.Driver.DownArrow}",
				Enumerable.Range (1, 5).Select (
						(a) => CreateMenuItem (a))
					.ToArray ());

			var mb = new MenuBar (new [] { menu }) {
				CanFocus = true,
				Width = 1,
				Y = Pos.Top (tf),
				X = Pos.Right (tf)
			};

			// HACKS required to make this work:
			mb.Enter += (s, e) => {
				// BUG: This does not select menu item 0
				// Instead what happens is the first keystroke the user presses
				// gets swallowed and focus is moved to 0.  Result is that you have
				// to press down arrow twice to select first menu item and/or have to
				// press Tab twice to move focus back to TextField
				mb.OpenMenu ();
			};

			Application.Top.Add (mb);
```

This mostly works today as @tznind notes. 

![4Sl96Je 1](https://user-images.githubusercontent.com/585482/233849864-b2666b06-8106-4821-a1da-5d083cbce0c6.gif)

What needs to be fixed to make this really be done:

- The menu should pop down aligned to the left of the textField. 
- MenuBorderStyle = LineStyle.None might look better

**Comment by Serg-Norseman** [comment]
After removing this component, how do I build an interface that requires functionality like selecting an existing element or entering new text? A ComboBox was the best choice here.

**Comment by tig** [comment]
#4412 will fix this.

There will probably be a 'DropDown <TView>' as a result. TView could be ListView or Menu or ColorPicker or TextField or....

Think of it as a non-modal version of Prompt that is rooted on a TextField (that can be read only or not). 

Sorry about the gap for now.

**Comment by Serg-Norseman** [comment]
TView - as I understand it, it will be the drop-down list? And the main part - the text field - will not be editable? 

So, if I need to use a classic ComboBox - an editable input field plus a drop-down list of static items - this won't work?

**Comment by tznind** [comment]
> TView - as I understand it, it will be the drop-down list? And the main part - the text field - will not be editable? 
> 
> So, if I need to use a classic ComboBox - an editable input field plus a drop-down list of static items - this won't work?


The proposal is for a normal text field which is fully editable.  And wiring to a popdown menu for autocomplete.

This would support
- edittable combo box use case
- setting for requiring exact match or not
Etc

**Comment by tig** [comment]
Correct. The twist I'm suggesting by having it be a generic with TView is the dropdown can be any view subclass. Eg a ColorPicker, OptionSelector, etc. 

We'll prob have a non generic thar defaults to ListView for the most common case.

**Comment by tig** [comment]
@tznind @BDisp and other subscribers:

Please review the spec that is now the first comment on this Issue and comment asap!

Thanks!

**Comment by BDisp** [comment]
This will be a great innovation for `PopoverMenu`, repurposing its functionality for a multitude of uses. I agree with all the returns listed in the table except for `ListView`. For "Autocomplete / combo-box," it should indeed internally return a string, but normally a developer needs to obtain the index from the `ListView` source to link to other objects. Of course, this can be provided through a property of the respective implementation derivation.

**Comment by tig** [comment]
Fixed
